src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs

```diff
@@ -4,42 +4,57 @@
     /// </summary>
     /// <seealso>https://aka.ms/msal-net-token-cache-serialization</seealso>
     public class MsalDistributedTokenCacheAdapter : MsalAbstractTokenCacheProvider
     {
         /// <summary>
         /// .NET Core Memory cache.
         /// </summary>
-        private readonly IDistributedCache _distributedCache;
+        internal /*for tests*/ readonly IDistributedCache _distributedCache;
+        internal /*for tests*/ readonly MemoryCache _memoryCache;
+        private readonly ILogger<MsalDistributedTokenCacheAdapter> _logger;
+        private readonly TimeSpan? _expirationTime;
 
         /// <summary>
-        /// MSAL memory token cache options.
+        /// MSAL distributed token cache options.
         /// </summary>
-        private readonly MsalDistributedTokenCacheAdapterOptions _cacheOptions;
+        private readonly MsalDistributedTokenCacheAdapterOptions _distributedCacheOptions;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="MsalDistributedTokenCacheAdapter"/> class.
         /// </summary>
-        /// <param name="memoryCache">Distributed cache instance to use.</param>
-        /// <param name="cacheOptions">Options for the token cache.</param>
+        /// <param name="distributedCache">Distributed cache instance to use.</param>
+        /// <param name="distributedCacheOptions">Options for the token cache.</param>
+        /// <param name="logger">MsalDistributedTokenCacheAdapter logger.</param>
         public MsalDistributedTokenCacheAdapter(
-                                            IDistributedCache memoryCache,
-                                            IOptions<MsalDistributedTokenCacheAdapterOptions> cacheOptions)
+                                            IDistributedCache distributedCache,
+                                            IOptions<MsalDistributedTokenCacheAdapterOptions> distributedCacheOptions,
+                                            ILogger<MsalDistributedTokenCacheAdapter> logger)
         {
-            if (cacheOptions == null)
+            if (distributedCacheOptions == null)
             {
-                throw new ArgumentNullException(nameof(cacheOptions));
+                throw new ArgumentNullException(nameof(distributedCacheOptions));
             }
 
-            _distributedCache = memoryCache;
-            _cacheOptions = cacheOptions.Value;
+            _distributedCache = distributedCache;
+            _distributedCacheOptions = distributedCacheOptions.Value;
+            _memoryCache = new MemoryCache(new MemoryCacheOptions { SizeLimit = _distributedCacheOptions.L1CacheSizeLimit * (1024 * 1024) });
+            _logger = logger;
+
+            if (_distributedCacheOptions.AbsoluteExpirationRelativeToNow != null)
+            {
+                _expirationTime = TimeSpan.FromMilliseconds(_distributedCacheOptions.AbsoluteExpirationRelativeToNow.Value.TotalMilliseconds * _distributedCacheOptions.L1ExpirationTimeRatio);
```

Comment:
Is it possible that the L1ExpirationTimeRatio is set to a number > 1? Should there be a validation?

Reply:
fixed. nice catch.

---
In reply to: [586317374](https://github.com/AzureAD/microsoft-identity-web/pull/1037#discussion_r586317374) [](ancestors = 586317374)